File: day1/solution.js
const fs = require('fs')

Array.prototype.sum = function () {
  return this.reduce((a, b) => a + b)
}

function processInput() {
  const raw_data = fs.readFileSync('./day1/input.txt', 'utf8');
  return raw_data
    .split('\n\n')
    .map(line => line.split('\n').map(Number))
}

const caloryListsPerElf = processInput()
const totalCaloriesPerElf = caloryListsPerElf.map(calories => calories.sum())

function solvePart1() {
  return Math.max(...totalCaloriesPerElf)
}

function solvePart2() {
  totalCaloriesPerElf.sort((a, b) => b - a)
  return totalCaloriesPerElf.slice(0, 3).sum()
}

console.log(solvePart1() == 68787)
console.log(solvePart2() == 198041)

File: day1/solution.scala
import scala.io.Source
import scala.math._

object Solution extends App {
  val nums = Source.fromFile("./input.txt").getLines.toList

  def sumCaloriesPerElf(calories: List[String], currSum: Int = 0, numLists: List[Int] = List.empty): List[Int] = calories match {
      case Nil => numLists
      case head :: tail if head.isEmpty => sumCaloriesPerElf(tail, currSum = 0, numLists :+ currSum)
      case head :: tail => sumCaloriesPerElf(tail, currSum + head.toInt, numLists)
  }

  def solvePart1(nums: List[String]): Int = sumCaloriesPerElf(nums).max

  def solvePart2(nums: List[String]): Int = {
    sumCaloriesPerElf(nums)
      .sorted(Ordering.Int.reverse)
      .take(3)
      .sum
  }

  println(solvePart1(nums) == 68787)
  println(solvePart2(nums) == 198041)
}
